spring-pulsar/src/main/java/org/springframework/pulsar/core/DefaultPulsarConsumerFactory.java

```diff
@@ -38,82 +41,173 @@
+	@Override
+	public Consumer<T> createConsumer(Schema<T> schema, List<ConsumerBuilderCustomizer<T>> customizers)
+			throws PulsarClientException {
 
-		if (!CollectionUtils.isEmpty(properties)) {
-			consumerBuilder.loadConf(properties);
-		}
-		Consumer<T> consumer = consumerBuilder.subscribe();
-		this.consumers.add(consumer);
-		return consumer;
+		return createConsumer(schema, null, null, Collections.emptyList());
 	}
 
 	@Override
-	public Consumer<T> createConsumer(Schema<T> schema, BatchReceivePolicy batchReceivePolicy,
-			Map<String, Object> propertiesToOverride) throws PulsarClientException {
-
+	public Consumer<T> createConsumer(Schema<T> schema, @Nullable Collection<String> topics,
+			@Nullable Map<String, String> properties, @Nullable List<ConsumerBuilderCustomizer<T>> customizers)
+			throws PulsarClientException {
 		final ConsumerBuilder<T> consumerBuilder = this.pulsarClient.newConsumer(schema);
-		final Map<String, Object> properties = new HashMap<>(this.consumerConfig);
-		properties.putAll(propertiesToOverride);
 
-		// Remove deadLetterPolicy from the properties here and save it to re-apply after
-		// calling `loadConf` (https://github.com/apache/pulsar/issues/11646)
-		DeadLetterPolicy deadLetterPolicy = null;
-		if (properties.containsKey("deadLetterPolicy")) {
-			deadLetterPolicy = (DeadLetterPolicy) properties.remove("deadLetterPolicy");
+		ConsumerConfigurationData<T> config = this.consumerConfig.clone();
+		if (topics != null) {
+			config.setTopicNames(new HashSet<>(topics));
 		}
-
-		if (!CollectionUtils.isEmpty(properties)) {
-			consumerBuilder.loadConf(properties);
-		}
-
-		if (deadLetterPolicy != null) {
-			consumerBuilder.deadLetterPolicy(deadLetterPolicy);
+		if (properties != null) {
+			config.setProperties(new TreeMap<>(properties));
 		}
 
-		if (properties.containsKey("negativeAckRedeliveryBackoff")) {
-			final RedeliveryBackoff negativeAckRedeliveryBackoff = (RedeliveryBackoff) properties
-					.get("negativeAckRedeliveryBackoff");
-			consumerBuilder.negativeAckRedeliveryBackoff(negativeAckRedeliveryBackoff);
-		}
+		loadConf(consumerBuilder, config);
 
-		if (properties.containsKey("ackTimeoutRedeliveryBackoff")) {
-			final RedeliveryBackoff ackTimeoutRedeliveryBackoff = (RedeliveryBackoff) properties
-					.get("ackTimeoutRedeliveryBackoff");
-			consumerBuilder.ackTimeoutRedeliveryBackoff(ackTimeoutRedeliveryBackoff);
+		if (!CollectionUtils.isEmpty(customizers)) {
+			customizers.forEach(customizer -> customizer.customize(consumerBuilder));
 		}
 
-		consumerBuilder.batchReceivePolicy(batchReceivePolicy);
 		Consumer<T> consumer = consumerBuilder.subscribe();
 		this.consumers.add(consumer);
 		return consumer;
 	}
 
-	public Map<String, Object> getConsumerConfig() {
+	public ConsumerConfigurationData<T> getConsumerConfig() {
 		return this.consumerConfig;
 	}
 
+	private static <T> void loadConf(ConsumerBuilder<T> consumer, ConsumerConfigurationData<T> conf) {
```

Comment:
> You're totally right that `ConsumerConfigurationData` is part of the client impl so it probably shouldn't be used outside of the pulsar-client internal. There used to be a `ConsumerConfiguration` in the API but it's deprecated. I see 3 options: 1/ Create a conf structure in spring-pulsar. It could have fluent setters and optional values. Mapping code would still be needed. 2/ Use a `ConsumerBuilderCustomizer` in the `DefaultPulsarConsumerFactory` instead of `ConsumerConfigurationData`. And also give the possibility to set default `topicNames` and `properties` that can be overriden later in `createConsumer`. 3/ Create a `DefaultPulsarConsumerFactoryBuilder` that can take set every prop individually.
> 
> My preference goes to option 2/

@cbornet 😞 Bummer. Thx for checking into that though. 

Of the above options I would go for 2) as well. I think let's get consensus on the signatures for `DefaultPulsarConsumerFactory.<init>` and `PulsarConsumerFactory` before moving forward though. 

There is also another option, leave the construction of DPCF using a map. I believe the primary use case is that the auto-configured DPCF will be used. Also, the auto-configuration hides this from the end user (`PulsarProperties->Map->DPCF.<init>`. So the type-safe `<init>` is not as big of a win as the type-safe `createConsumer` method w/ use of the customizers.